### ISSUE TITLE
Missing AccessorDeclarationSyntax.ExpressionBody

### ISSUE THREAD

**Issue by Logerfo** [dotnet/roslyn#19404]
I'm creating an `CodeFixProvider` for an analyzer and stepped on the following issue:

According to the [source ](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Generated/Syntax.xml.Internal.Generated.cs,22db92ca60784def), `AccessorDeclarationSyntax.ExpressionBody` is `public`. I can see that property through VS watch, but I can't access it through my normal code ('AccessorDeclarationSnytax' does not contain a definition for 'ExpressionBody') - note that this is not an "inaccessible due to its protection level" error. As an expression bodied property accessor contains no `Body` (property is `null`), I can't be sure if whether my property is an "auto-prop" or not. Am I missing something here?

Please refer me to the right place to ask this if it's not here.

**Comment by Logerfo** [comment]
I was able to find (or not) an `ArrowExpressionClauseSyntax` through `SyntaxNode.ChildNodes()`, but still, why does `ExpressionBody` appear on VS watch but my code can't see it?

**Comment by jcouv** [comment]
@Logerfo, 
Could you confirm the fully-qualified type name that you can see in the QuickWatch window? There are two kinds of `AccessorDeclarationSyntax` (one is internal and the other is public). I expect you should only see the public type, but I just want to confirm.

**Comment by Logerfo** [comment]
@jcouv QuickWatch shows the following expression:
`(new System.Linq.SystemCore_EnumerableDebugView<Microsoft.CodeAnalysis.CSharp.Syntax.AccessorDeclarationSyntax>(declaration.AccessorList.Accessors).Items[0]).ExpressionBody`
So, to answer your question: `Microsoft.CodeAnalysis.CSharp.Syntax.AccessorDeclarationSyntax`

**Comment by jcouv** [comment]
Thanks for the clarification. This is indeed the public `AccessorDeclarationSyntax`, so I have no idea why the property would give you an error for missing member `ExpressionBody`.

@gafter Do you have some clue how to approach this issue? I'm grasping at straws.

@Logerfo If you don't mind, another "straw" that may be worth checking is the version of the CodeAnalysis and CodeAnalysis.CSharp that you're using.
![image](https://cloud.githubusercontent.com/assets/12466233/26009658/a69caf80-36fe-11e7-8f4f-fa975ce25c9c.png)

**Comment by Logerfo** [comment]
@jcouv what window is that? 

I don't know if this is relevant, but here's the thing with versions: as soon as I created the analyzer solution, nuget packages were outdated. `Common` and `CSharp` both are 1.0.1, but 2.1.0 is available. However, I can't update them:
>Could not install package 'Microsoft.CodeAnalysis.Common 2.1.0'. You are trying to install this package into a project that targets '.NETPortable,Version=v4.5,Profile=Profile7', but the package does not contain any assembly references or content files that are compatible with that framework.

So I thought "what if I bump the .NET version?". Then I went to the analyzer project properties > targeting > change > .NET 4.6 > OK:
>The project's targets cannot be changed. The selected targets require the project to opt-into NuGet 3.0 support, however, Visual Studio cannot automatically do this for you. Please uninstall all NuGet packages and try again.

This is the first time I'm playing with analyzers, therefore I have no idea what's happening :D

**Comment by jcouv** [comment]
I see. The fact that CodeAnalysis was 1.0.1 likely explains why `ExpressionBody` was not available.
The 2.0 compiler requires .NET 4.6, so you are on the right track.

As far as the project and nuget issue, I'm not sure. @sharwell may be able to assist.

**Comment by Logerfo** [comment]
@jcouv if it is a version problem, why does the property shows up in the QuickWatch? Shouldn't it use the same version I've used to compile the analyzer?

**Comment by jcouv** [comment]
I don't think the QuickWatch window respects accessibility (it even shows private fields, for instance). It's possible the property existed, but was not yet public.
To be completely sure what is going on, you'd want to find the filepath for CodeAnalysis binaries (as shown above in Modules window), open the CodeAnalysis.CSharp binary in `ildasm` or `ILSpy` and check the accessibility the property had back then.

**Comment by Logerfo** [comment]
```CIL
.method public hidebysig specialname instance class Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.ArrowExpressionClauseSyntax 
        get_ExpressionBody() cil managed
{
  // Code size       7 (0x7)
  .maxstack  8
  IL_0000:  ldarg.0
  IL_0001:  ldfld      class Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.ArrowExpressionClauseSyntax Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.AccessorDeclarationSyntax::expressionBody
  IL_0006:  ret
} // end of method AccessorDeclarationSyntax::get_ExpressionBody
```
@jcouv  Does that mean anything to you? Seems like hieroglyph for me

**Comment by jcouv** [comment]
Yes, `public` is the key there.

Looking up the thread, I see something weird. It seems you're not trying to access `ExpressionBody` on a `AccessorDeclarationSyntax`, but on a `SystemCore_EnumerableDebugView<...>`:
`(new System.Linq.SystemCore_EnumerableDebugView<Microsoft.CodeAnalysis.CSharp.Syntax.AccessorDeclarationSyntax>(declaration.AccessorList.Accessors).Items[0]).ExpressionBody`

What is the full error message you get for missing member?

**Comment by Logerfo** [comment]
CS1061	'AccessorDeclarationSyntax' does not contain a definition for 'ExpressionBody' and no extension method 'ExpressionBody' accepting a first argument of type 'AccessorDeclarationSyntax' could be found (are you missing a using directive or an assembly reference?)

**Comment by sharwell** [comment]
> if it is a version-related problem, why does the property shows up in the QuickWatch? Shouldn't it use the same version I've used to compile the analyzer?

Assembly binding redirection is used to make your analyzers operate against the most recent known version at runtime.

* If you compile against the same version as the IDE uses, you'll have a version match
* If you compile against an older version than the IDE uses, you'll see the newer version at runtime (and while debugging). The API you see when you compile will be a subset of the API that actually exists on the objects at runtime.
* If you compile against a newer version than the IDE uses, your analyzer will not work at runtime.

By compiling against Roslyn 1.0.1, you fall into the second point for everything newer than Visual Studio 2015 RTM. With this selection, your analyzer is compatible with the most versions of Visual Studio possible for Roslyn, but you will not be able to specifically analyze C# 7 syntax (because they were added in the Roslyn 2.x API).

> As far as the project and nuget issue, I'm not sure. @sharwell may be able to assist.

When I last tried this migration, I kept hitting issues (which may now be resolved). A pull request was submitted to StyleCop Analyzers which claims to make the change you are requesting, but I haven't yet reviewed it in detail: DotNetAnalyzers/StyleCopAnalyzers#2327

**Comment by Logerfo** [comment]
@sharwell Thanks for replying. Am I supposed to wait until VS RTM gets this to-be-merged PR then?

**Comment by sharwell** [comment]
@Logerfo That PR is a third-party project trying to accomplish what you are trying to accomplish (update from using 1.x Roslyn to using 2.x Roslyn). Presumably, if you take the same steps in your project as demonstrated in that PR, you will be able to start working with the Roslyn 2.x API.

**Comment by Logerfo** [comment]
@sharwell Well, I tried, but couldn't manage to do that. Should I expect anything from the IDE to be released? A migration tool or project creation with the lastest roslyn version?

**Comment by Logerfo** [comment]
According to https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/2327#issuecomment-307642203, the migration was superseded by https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/2370, which uses reflection to find members like the one I mentioned, as seen here https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/2370/commits/00da27631f2a228a817e20befba4839ae18b61ce#diff-29450f672610c72de247ff6ba3a1b7ceR22. That was done in order to maintain support to VS2015, which is not my case, but I couldn't manage to succeed at migrating.

**Comment by sharwell** [comment]
@Logerfo Yes, we definitely took things in a different direction for that project. I'm really excited about the possibilities for supporting both Visual Studio 2015 users and C# 7 users from a single library.

**Comment by sharwell** [comment]
@rchande Do you know if some of the recent dependencies work makes the NuGet packages easier to install in external projects with less "friction"?

**Comment by CyrusNajmabadi** [comment]
Closing out as this is stale.
